File: my-library/Chinese_Remainder_Theorem.py
def gcd_ext(a, b):
    if b == 0:
        return 1, 0, a
    else:
        y, x, res = gcd_ext(b, a % b)
        y -= a // b * x
        return x, y, res


def chreth(n, a, m, b):
    k1, k2, d = gcd_ext(n, m)
    if (b - a) % d != 0:
        return -1
    return (k1 * (b - a) // d * n + a) % (n * m // d)


print(chreth(int(input()), int(input()), int(input()), int(input())))

File: my-library/Math/Chinese Remainder Theorem.cpp
#include <iostream>

using namespace std;

int gcd_ext(int a, int b, int &x, int &y) {
  if (b == 0) {
    x = 1;
    y = 0;
    return a;
  } else {
    int res = gcd_ext(b, a % b, y, x);
    y -= a / b * x;
    return res;
  }
}

int chreth(int n, int a, int m, int b) {
  int k1, k2;
  int d = gcd_ext(n, m, k1, k2);
  if ((b - a) % d != 0) {
    return -1;
  }
  int mod = n * m / d;
  int x = ((k1 * (b - a) / d * n + a) % mod + mod) % mod;
  return x;
}
